multimodal/src/autogluon/multimodal/models/sam.py

```diff
@@ -59,16 +323,37 @@ def __init__(
         freeze_model_layers(self.model, self.frozen_layers)
 
         self.image_size = self.model.vision_encoder.image_size
+        self.config = self.model.config
 
+        self.model.mask_decoder.num_mask_tokens = num_mask_tokens
         # for Binary Semantic Segmentation Tasks.
-        self.model.mask_decoder.num_mask_tokens = 1
-        mask_token_data = self.model.mask_decoder.mask_tokens.weight.data[0]
-        self.model.mask_decoder.mask_tokens = nn.Embedding(1, self.model.mask_decoder.hidden_size)
-        self.model.mask_decoder.mask_tokens.weight.data[0] = mask_token_data
-        hyper_mlps = self.model.mask_decoder.output_hypernetworks_mlps[0]
-        self.model.mask_decoder.output_hypernetworks_mlps = nn.ModuleList([hyper_mlps])
+        if num_classes == 1:
+            mask_token_data = self.model.mask_decoder.mask_tokens.weight.data[0]
+            self.model.mask_decoder.mask_tokens = nn.Embedding(1, self.model.mask_decoder.hidden_size)
+            self.model.mask_decoder.mask_tokens.weight.data[0] = mask_token_data
+            hyper_mlps = self.model.mask_decoder.output_hypernetworks_mlps[0]
+            self.model.mask_decoder.output_hypernetworks_mlps = nn.ModuleList([hyper_mlps])
+        # for Multi-class Semantic Segmentation Tasks.
+        else:
+            self.model.mask_decoder.num_classes = num_classes
+            mask_token_data = self.model.mask_decoder.mask_tokens.weight.data[0]
+            self.model.mask_decoder.mask_tokens = nn.Embedding(num_mask_tokens, self.model.mask_decoder.hidden_size)
+            for i in range(num_mask_tokens):
+                self.model.mask_decoder.mask_tokens.weight.data[i] = mask_token_data
 
-        self.config = self.model.config
+            hyper_mlps = self.model.mask_decoder.output_hypernetworks_mlps[0]
+            self.model.mask_decoder.output_hypernetworks_mlps = nn.ModuleList([hyper_mlps])
+            self.model.mask_decoder.output_classifier_mlps = nn.Linear(
+                self.model.mask_decoder.hidden_size, num_classes + 1
+            )
```

Comment:
Some code is repeated in both binary and multiclass. Need to reorganize the code reflect the differences between binary and multiclass better.